### PR TITLE
core: add block init with all the fields

### PR DESCRIPTION
**Pull request by superlopuh**
This gives an alternative to the builder methods. A subsequent PR will deprecate the other buidlers and migrate existing code to the API we settle on.

One alternative to consider is passing a list of ops as opposed to a vararg. I'm happy with either.

### CODE REVIEW

**Review by georgebisbas — COMMENTED** [review]
What is the motivation behind this PR?

**Review by webmiche — APPROVED** [review]
LGTM, seems cleaner than the old API :+1:

**Comment by superlopuh** [comment]
@georgebisbas mostly cleanup, I'm slowly trying to migrate everything that makes sense to inits rather than builder methods. I find having a single API for object creation conceptually simpler.

**Comment by superlopuh** [comment]
Actually on second thought, I think taking an Iterable reflects the fact that it's a container of a list better, I changed the init. @webmiche does it still look ok to you?

**Comment by webmiche** [comment]
Yes, still LGTM. I even think its a bit better now, varargs have an implicit taste in my mouth :wink:

**Review by math-fehr — APPROVED** [review]
:+1:

File: xdsl/ir.py [hunk 1]
```diff
@@ -855,22 +855,36 @@ def irdl_definition(cls) -> OpDef | None:
...
+    ops: list[Operation]
     """Ordered operations contained in the block."""
 
-    parent: Region | None = field(default=None, init=False, repr=False)
+    parent: Region | None
     """Parent region containing the block."""
 
+    def __init__(self,
+                 ops: Iterable[Operation] = (),
+                 arg_types: tuple[Attribute, ...] = (),
```
Comment (math-fehr) [inline xdsl/ir.py:875]: Can we remain consistent with `Region` and make `parent` and `declared_at` kw_only?

**Pull request merged**